File: ndfs-py/src/ndfs/master_block.py
"""
NDFS master block: the 32-byte structure at offset 2016 of page 0.
Also handles the 16-byte extended info block at offset 2000.

SPDX-License-Identifier: MIT
Copyright (c) 1985-2026 Ronny Hansen
HackerCorp Labs - https://github.com/HackerCorpLabs
"""
from __future__ import annotations

from typing import Optional, Union

from ndfs.block_pointer import BlockPointer
from ndfs.constants import (
    MASTER_BLOCK_OFFSET,
    EXTENDED_INFO_OFFSET,
    NDFS_PAGE_SIZE,
    NDFS_NAME_MAX,
)
from ndfs.types import ChecksumValidation
from ndfs.endian import read_uint16_be, read_uint32_be, write_uint16_be, write_uint32_be
from ndfs.ndfs_name import read_ndfs_name, write_ndfs_name

_BufType = Union[bytes, bytearray, memoryview]


class MasterBlock:
    """The NDFS master block and extended info parsed from page 0."""

    __slots__ = (
        "directory_name",
        "object_file_pointer",
        "user_file_pointer",
        "bit_file_pointer",
        "unreserved_pages",
        "image_size",
        "ext_checksum",
        "ext_reserved1",
        "ext_reserved2",
        "ext_reserved3",
        "ext_flag_word",
        "ext_last_system_number",
        "ext_pages_available",
        "ext_calculated_checksum",
        "ext_valid",
        "checksum_state",
        "has_flomon",
    )

    def __init__(self) -> None:
        self.directory_name: str = ""
        self.object_file_pointer: Optional[BlockPointer] = None
        self.user_file_pointer: Optional[BlockPointer] = None
        self.bit_file_pointer: Optional[BlockPointer] = None
        self.unreserved_pages: int = 0
        self.image_size: int = 0

        # Extended info fields
        self.ext_checksum: int = 0
        self.ext_reserved1: int = 0
        self.ext_reserved2: int = 0
        self.ext_reserved3: int = 0
        self.ext_flag_word: int = 0
        self.ext_last_system_number: int = 0
        self.ext_pages_available: int = 0
        self.ext_calculated_checksum: int = 0
        self.ext_valid: bool = False
        self.checksum_state: ChecksumValidation = ChecksumValidation.Invalid
        self.has_flomon: bool = False

    #: Extended-info flag word (1754B) bit 15: "directory entered / in use".
    #:
    #: The kernel sets it on enter (CHDSI 37763B, ``BSET ONE 170``) and clears it on
    #: release (REENB 40162B, ``BSET ZRO 170``). A stored 0x8000 means the volume was left
    #: entered by the system in ``ext_last_system_number``. Bits 0-14 have no known
    #: meaning -- preserve them verbatim.
    #:
    #: NOTE: system number 0 means NO OWNER recorded, not "system zero".
    EXT_FLAG_ENTERED = 0x8000

    @property
    def is_directory_entered(self) -> bool:
        """True when the volume was left 'entered' (flag bit 15) and not cleanly released."""
        return (self.ext_flag_word & MasterBlock.EXT_FLAG_ENTERED) != 0

    @staticmethod
    def compute_ext_checksum(
        reserved1: int, reserved2: int, reserved3: int,
        flag_word: int, system_number: int,
        pages_hi: int, pages_lo: int,
    ) -> int:
        """Compute the extended-info checksum exactly as the SINTRAN kernel does.

        A plain 16-bit ADDITIVE SUM of the seven words that FOLLOW the checksum slot
        (words 1751B-1757B), truncated to 16 bits.

        Kernel-proven from both sides of the carved 006-S3FS segment: the writer WXDIR
        (37702B) and the enter-directory validator CHDSI (37763B) run the identical
        ``ADD ,X 0`` accumulation loop. There is no XOR, and the system number is simply
        one of the seven summed words -- not a separate addend.
        """
        total = (reserved1 + reserved2 + reserved3
                 + flag_word + system_number + pages_hi + pages_lo)
        return total & 0xFFFF

    @classmethod
    def from_bytes(cls, page_data: _BufType) -> MasterBlock:
        """Parse a master block (and extended info) from a full page 0 buffer.

        The buffer must be at least NDFS_PAGE_SIZE bytes.
        """
        if len(page_data) < NDFS_PAGE_SIZE:
            raise ValueError("Page data too small for master block")

        mb = cls()
        off = MASTER_BLOCK_OFFSET

        # Directory name (16 bytes, terminated by 0x27)
        mb.directory_name = read_ndfs_name(page_data, off, NDFS_NAME_MAX)

        # Block pointers
        mb.object_file_pointer = BlockPointer.from_bytes(page_data, off + 0x10)
        mb.user_file_pointer = BlockPointer.from_bytes(page_data, off + 0x14)
        mb.bit_file_pointer = BlockPointer.from_bytes(page_data, off + 0x18)

        # Unreserved pages
        mb.unreserved_pages = read_uint32_be(page_data, off + 0x1C)

        # --- Extended info (bytes 2000-2015) ---
        ext = EXTENDED_INFO_OFFSET
        mb.ext_checksum = read_uint16_be(page_data, ext)
        mb.ext_reserved1 = read_uint16_be(page_data, ext + 2)
        mb.ext_reserved2 = read_uint16_be(page_data, ext + 4)
        mb.ext_reserved3 = read_uint16_be(page_data, ext + 6)
        mb.ext_flag_word = read_uint16_be(page_data, ext + 8)
        mb.ext_last_system_number = read_uint16_be(page_data, ext + 10)
        mb.ext_pages_available = read_uint32_be(page_data, ext + 12)

        # KERNEL-CORRECTED: the checksum is a plain 16-bit ADDITIVE SUM of the seven words
        # FOLLOWING it (1751B-1757B) -- NOT "XOR six words, then add the system number".
        # Proven from both sides of the real SINTRAN kernel (carved 006-S3FS): the writer
        # WXDIR (37702B) and the enter-directory validator CHDSI (37763B) run the identical
        # `ADD ,X 0` accumulation loop. The XOR form matched PACK-ONE only by coincidence
        # (its only two words sharing a set bit are flag=0x8000 and pages_lo=0x9051, both
        # bit 15 -- they cancel under XOR and carry out under ADD, giving the same low 16
        # bits). Verified against three real disks: 0x10B7, 0x1051, 0xC162.
        pages_lo = mb.ext_pages_available & 0xFFFF
        pages_hi = (mb.ext_pages_available >> 16) & 0xFFFF
        calculated = MasterBlock.compute_ext_checksum(
            mb.ext_reserved1, mb.ext_reserved2, mb.ext_reserved3,
            mb.ext_flag_word, mb.ext_last_system_number, pages_hi, pages_lo,
        )
        mb.ext_calculated_checksum = calculated

        # The kernel writes and compares the FULL 16 bits, so there is no "low byte only"
        # accept state -- that was a reader-side heuristic SINTRAN never produces.
        mb.checksum_state = (
            ChecksumValidation.Valid
            if mb.ext_checksum == calculated
            else ChecksumValidation.Invalid
        )

        # Detect FLOMON
        mb.has_flomon = MasterBlock._detect_flomon(page_data)

        # Extended info validity
        if mb.has_flomon:
            mb.ext_valid = False
        else:
            checksum_non_zero = mb.ext_checksum != 0
            checksum_ok = mb.checksum_state == ChecksumValidation.Valid
            mb.ext_valid = checksum_non_zero and checksum_ok

        return mb

    @staticmethod
    def _detect_flomon(page_data: _BufType) -> bool:
        """Detect FLOMON boot format.

        FLOMON disks have a simplified boot loader where the BPUN binary section
        has address=0, count=0, and checksum=0.
        """
        # Look for '!' marker in the first part of the page
        exclamation_pos = -1
        limit = min(len(page_data), 256)
        for i in range(limit):
            if page_data[i] == 0x21:
                exclamation_pos = i
                break
        if exclamation_pos < 0:
            return False

        # After '!', BPUN has: address(2) + count(2) bytes
        after_excl = exclamation_pos + 1
        if after_excl + 4 > len(page_data):
            return False

        addr = read_uint16_be(page_data, after_excl)
        count = read_uint16_be(page_data, after_excl + 2)

        return addr == 0 and count == 0

    def is_valid(self) -> bool:
        """Check if the master block is valid."""
        # Check directory name is printable ASCII
        if len(self.directory_name) > 0:
            for ch in self.directory_name:
                c = ord(ch)
                if c < 0x20 or c > 0x7E:
                    return False

        # At least one pointer must be valid, or a directory name must exist
        has_valid_pointer = False
        if self.object_file_pointer is not None and self.object_file_pointer.is_valid():
            has_valid_pointer = True
        if self.user_file_pointer is not None and self.user_file_pointer.is_valid():
            has_valid_pointer = True
        if self.bit_file_pointer is not None and self.bit_file_pointer.is_valid():
            has_valid_pointer = True

        return has_valid_pointer or len(self.directory_name) > 0

    def write_to_bytes(self, page_data: bytearray) -> None:
        """Write the master block to page data at the standard offset."""
        if len(page_data) < NDFS_PAGE_SIZE:
            raise ValueError("Page buffer too small for master block")

        off = MASTER_BLOCK_OFFSET

        # Clear the master block area
        for i in range(32):
            page_data[off + i] = 0

        # Directory name
        write_ndfs_name(page_data, off, self.directory_name, NDFS_NAME_MAX)

        # Block pointers
        if self.object_file_pointer is not None:
            self.object_file_pointer.to_bytes(page_data, off + 0x10)
        if self.user_file_pointer is not None:
            self.user_file_pointer.to_bytes(page_data, off + 0x14)
        if self.bit_file_pointer is not None:
            self.bit_file_pointer.to_bytes(page_data, off + 0x18)

        # Unreserved pages
        write_uint32_be(page_data, off + 0x1C, self.unreserved_pages)

    def write_extended_info(self, page_data: bytearray) -> None:
        """Write extended info to page data."""
        if len(page_data) < NDFS_PAGE_SIZE:
            raise ValueError("Page buffer too small for extended info")

        ext = EXTENDED_INFO_OFFSET

        # Calculate checksum
        pages_lo = self.ext_pages_available & 0xFFFF
        pages_hi = (self.ext_pages_available >> 16) & 0xFFFF
        # Recompute the checksum from the current fields, exactly as the kernel writer
        # WXDIR (37702B) does: a 16-bit additive sum of words 1751B-1757B. Every write of
        # the extended-info block is followed by a fresh checksum.
        checksum = MasterBlock.compute_ext_checksum(
            self.ext_reserved1, self.ext_reserved2, self.ext_reserved3,
            self.ext_flag_word, self.ext_last_system_number, pages_hi, pages_lo,
        )

        write_uint16_be(page_data, ext, checksum)
        write_uint16_be(page_data, ext + 2, self.ext_reserved1)
        write_uint16_be(page_data, ext + 4, self.ext_reserved2)
        write_uint16_be(page_data, ext + 6, self.ext_reserved3)
        write_uint16_be(page_data, ext + 8, self.ext_flag_word)
        write_uint16_be(page_data, ext + 10, self.ext_last_system_number)
        write_uint32_be(page_data, ext + 12, self.ext_pages_available)

File: ndfs-ts/src/master-block.ts
/**
 * NDFS master block: the 32-byte structure at offset 2016 of page 0.
 * Also handles the 16-byte extended info block at offset 2000.
 *
 * SPDX-License-Identifier: MIT
 * Copyright (c) 1985-2026 Ronny Hansen
 * HackerCorp Labs - https://github.com/HackerCorpLabs
 */

import { BlockPointer } from './block-pointer.js';
import {
  MASTER_BLOCK_OFFSET,
  EXTENDED_INFO_OFFSET,
  NDFS_PAGE_SIZE,
  NDFS_NAME_MAX,
} from './constants.js';
import { ChecksumValidation } from './types.js';
import { readUint16BE, readUint32BE, writeUint16BE, writeUint32BE } from './endian.js';
import { readNdfsName, writeNdfsName } from './ndfs-name.js';

export class MasterBlock {
  directoryName: string = '';
  objectFilePointer: BlockPointer | null = null;
  userFilePointer: BlockPointer | null = null;
  bitFilePointer: BlockPointer | null = null;
  unreservedPages: number = 0;
  imageSize: number = 0;

  // Extended info fields
  extChecksum: number = 0;
  extReserved1: number = 0;
  extReserved2: number = 0;
  extReserved3: number = 0;
  extFlagWord: number = 0;
  extLastSystemNumber: number = 0;
  extPagesAvailable: number = 0;
  extCalculatedChecksum: number = 0;
  extValid: boolean = false;
  checksumState: ChecksumValidation = ChecksumValidation.Invalid;
  hasFlomon: boolean = false;

  /**
   * Extended-info flag word (1754B) bit 15: "directory entered / in use".
   *
   * The kernel sets it on enter (CHDSI 37763B, `BSET ONE 170`) and clears it on release
   * (REENB 40162B, `BSET ZRO 170`). A stored 0x8000 means the volume was left entered by
   * the system in `extLastSystemNumber`. Bits 0-14 have no known meaning — preserve them.
   *
   * NOTE: system number 0 means NO OWNER recorded — never report "entered by system 0".
   */
  static readonly EXT_FLAG_ENTERED = 0x8000;

  /** True when the volume was left "entered" (flag bit 15 set) and not cleanly released. */
  get isDirectoryEntered(): boolean {
    return (this.extFlagWord & MasterBlock.EXT_FLAG_ENTERED) !== 0;
  }

  /**
   * Compute the extended-info checksum exactly as the kernel does: a plain 16-bit
   * ADDITIVE SUM of the seven words FOLLOWING the checksum slot (words 1751B-1757B).
   *
   * Kernel-proven from both sides of the carved 006-S3FS segment — the writer WXDIR
   * (37702B) and the enter-directory validator CHDSI (37763B) run the identical
   * `ADD ,X 0` accumulation loop. There is no XOR, and the system number is simply one of
   * the seven summed words, not a separate addend.
   */
  static computeExtChecksum(
    reserved1: number,
    reserved2: number,
    reserved3: number,
    flagWord: number,
    systemNumber: number,
    pagesHi: number,
    pagesLo: number,
  ): number {
    const sum =
      reserved1 + reserved2 + reserved3 + flagWord + systemNumber + pagesHi + pagesLo;
    return sum & 0xffff;
  }

  /**
   * Parse a master block (and extended info) from a full page 0 buffer.
   * The buffer must be at least NDFS_PAGE_SIZE bytes.
   */
  static fromBytes(pageData: Uint8Array): MasterBlock {
    if (pageData.length < NDFS_PAGE_SIZE) {
      throw new Error('Page data too small for master block');
    }

    const mb = new MasterBlock();
    const off = MASTER_BLOCK_OFFSET;

    // Directory name (16 bytes, terminated by 0x27)
    mb.directoryName = readNdfsName(pageData, off, NDFS_NAME_MAX);

    // Block pointers
    mb.objectFilePointer = BlockPointer.fromBytes(pageData, off + 0x10);
    mb.userFilePointer = BlockPointer.fromBytes(pageData, off + 0x14);
    mb.bitFilePointer = BlockPointer.fromBytes(pageData, off + 0x18);

    // Unreserved pages
    mb.unreservedPages = readUint32BE(pageData, off + 0x1c);

    // --- Extended info (bytes 2000-2015) ---
    const ext = EXTENDED_INFO_OFFSET;
    mb.extChecksum = readUint16BE(pageData, ext);
    mb.extReserved1 = readUint16BE(pageData, ext + 2);
    mb.extReserved2 = readUint16BE(pageData, ext + 4);
    mb.extReserved3 = readUint16BE(pageData, ext + 6);
    mb.extFlagWord = readUint16BE(pageData, ext + 8);
    mb.extLastSystemNumber = readUint16BE(pageData, ext + 10);
    mb.extPagesAvailable = readUint32BE(pageData, ext + 12);

    // KERNEL-CORRECTED: the checksum is a plain 16-bit ADDITIVE SUM of the seven words
    // FOLLOWING it (1751B-1757B) — NOT "XOR six words, then add the system number".
    // Proven from both sides of the real SINTRAN kernel (carved 006-S3FS): the writer
    // WXDIR (37702B) and the validator CHDSI (37763B) run the identical `ADD ,X 0` loop.
    // The XOR form matched PACK-ONE only by coincidence (its only two words sharing a set
    // bit are flag=0x8000 and pagesLo=0x9051, both bit 15 — which cancel under XOR and
    // carry out under ADD, giving the same low 16 bits). Verified on three real disks.
    const pagesLo = mb.extPagesAvailable & 0xffff;
    const pagesHi = (mb.extPagesAvailable >>> 16) & 0xffff;
    const calculated = MasterBlock.computeExtChecksum(
      mb.extReserved1,
      mb.extReserved2,
      mb.extReserved3,
      mb.extFlagWord,
      mb.extLastSystemNumber,
      pagesHi,
      pagesLo,
    );
    mb.extCalculatedChecksum = calculated;

    // The kernel writes and compares the FULL 16 bits, so there is no "low byte only"
    // accept state — that was a reader-side heuristic SINTRAN never produces.
    mb.checksumState =
      mb.extChecksum === calculated ? ChecksumValidation.Valid : ChecksumValidation.Invalid;

    // Detect FLOMON: all extended fields are zero (floppy)
    // FLOMON signature: address(2) + count(2) + checksum(2) all zero in boot area
    // Simplified: check if the first 6 bytes after boot offset look like FLOMON
    mb.hasFlomon = MasterBlock.detectFlomon(pageData);

    // Extended info validity
    if (mb.hasFlomon) {
      mb.extValid = false;
    } else {
      const checksumNonZero = mb.extChecksum !== 0;
      const checksumOk = mb.checksumState === ChecksumValidation.Valid;
      mb.extValid = checksumNonZero && checksumOk;
    }

    return mb;
  }

  /**
   * Detect FLOMON boot format by checking for the zero-address/count/checksum pattern.
   * FLOMON disks have a simplified boot loader where the BPUN binary section
   * has address=0, count=0, and checksum=0.
   */
  private static detectFlomon(pageData: Uint8Array): boolean {
    // Look for '!' marker in the first part of the page
    let exclamationPos = -1;
    for (let i = 0; i < Math.min(pageData.length, 256); i++) {
      if (pageData[i] === 0x21) {
        exclamationPos = i;
        break;
      }
    }
    if (exclamationPos < 0) return false;

    // After '!', BPUN has: address(2) + count(2) bytes
    // FLOMON has address=0, count=0
    const afterExcl = exclamationPos + 1;
    if (afterExcl + 4 > pageData.length) return false;

    const addr = readUint16BE(pageData, afterExcl);
    const count = readUint16BE(pageData, afterExcl + 2);

    return addr === 0 && count === 0;
  }

  /** Check if the master block is valid. */
  isValid(): boolean {
    // Check directory name is printable ASCII
    if (this.directoryName.length > 0) {
      for (let i = 0; i < this.directoryName.length; i++) {
        const c = this.directoryName.charCodeAt(i);
        if (c < 0x20 || c > 0x7e) return false;
      }
    }

    // At least one pointer must be valid, or a directory name must exist
    let hasValidPointer = false;
    if (this.objectFilePointer !== null && this.objectFilePointer.isValid()) hasValidPointer = true;
    if (this.userFilePointer !== null && this.userFilePointer.isValid()) hasValidPointer = true;
    if (this.bitFilePointer !== null && this.bitFilePointer.isValid()) hasValidPointer = true;

    return hasValidPointer || this.directoryName.length > 0;
  }

  /** Write the master block to page data at the standard offset. */
  writeToBytes(pageData: Uint8Array): void {
    if (pageData.length < NDFS_PAGE_SIZE) {
      throw new Error('Page buffer too small for master block');
    }

    const off = MASTER_BLOCK_OFFSET;

    // Clear the master block area
    pageData.fill(0, off, off + 32);

    // Directory name
    writeNdfsName(pageData, off, this.directoryName, NDFS_NAME_MAX);

    // Block pointers
    if (this.objectFilePointer) this.objectFilePointer.toBytes(pageData, off + 0x10);
    if (this.userFilePointer) this.userFilePointer.toBytes(pageData, off + 0x14);
    if (this.bitFilePointer) this.bitFilePointer.toBytes(pageData, off + 0x18);

    // Unreserved pages
    writeUint32BE(pageData, off + 0x1c, this.unreservedPages);
  }

  /** Write extended info to page data. */
  writeExtendedInfo(pageData: Uint8Array): void {
    if (pageData.length < NDFS_PAGE_SIZE) {
      throw new Error('Page buffer too small for extended info');
    }

    const ext = EXTENDED_INFO_OFFSET;

    // Recompute the checksum from the current fields, exactly as the kernel writer WXDIR
    // (37702B) does: a 16-bit additive sum of words 1751B-1757B. Every write of the
    // extended-info block is followed by a fresh checksum.
    const pagesLo = this.extPagesAvailable & 0xffff;
    const pagesHi = (this.extPagesAvailable >>> 16) & 0xffff;
    const checksum = MasterBlock.computeExtChecksum(
      this.extReserved1,
      this.extReserved2,
      this.extReserved3,
      this.extFlagWord,
      this.extLastSystemNumber,
      pagesHi,
      pagesLo,
    );

    writeUint16BE(pageData, ext, checksum);
    writeUint16BE(pageData, ext + 2, this.extReserved1);
    writeUint16BE(pageData, ext + 4, this.extReserved2);
    writeUint16BE(pageData, ext + 6, this.extReserved3);
    writeUint16BE(pageData, ext + 8, this.extFlagWord);
    writeUint16BE(pageData, ext + 10, this.extLastSystemNumber);
    writeUint32BE(pageData, ext + 12, this.extPagesAvailable);
  }
}
